---
author: "University of Melbourne Centre for Cancer Research"
date: "`r Sys.time()`"
output:
  html_document:
    theme: cosmo # darkly
    css: style.css
    toc: false
    code_download: true
  rmdformats::material:
    highlight: kate
params:
  title: "UMCCR Cancer Report"
  tumor_name: 'x'
  batch_name: 'x'
  genome_build: 'hg38' # 'hg19'
  key_genes: 'x'
  af_global: 'x'
  af_keygenes: 'x'
  somatic_snv: 'x'
  somatic_sv: 'x'
  somatic_sv_vcf: 'x'
  purple_som_gene_cnv: 'x'
  purple_som_cnv: 'x'
  purple_germ_cnv: 'x'
  purple_purity: 'x'
  purple_qc: 'x'
  purple_som_snv_vcf: 'x'
  oncoviral_present_viruses: 'x'
  oncoviral_breakpoints_tsv: 'x'
  conda_list: 'x'
  result_outdir: 'x'
description: "Analysis of tumor/normal samples at UMCCR"
title: "`r paste(params$batch_name, params$title)`"
---

```{r knitr_opts, include=F}
knitr::opts_chunk$set(collapse = TRUE, echo = FALSE,
                      warning = FALSE, message = FALSE)
```

```{r render_report_interactively, echo=F, eval=FALSE}
batch_name <- "SEQCII_50pc__T_SRR7890936_50pc"
# tumor_name <- strsplit(batch_name, "__")[[1]][2]
tumor_name <- "T_SRR7890936_50pc"
dd <- "/Users/pdiakumis/Desktop/projects/umccr/umccrise_dev/umccrise/umccrise/rmd_files"

params_tmp <- list(
  gadi = list(
    tumor_name='T_SRR7890936_50pc',
    batch_name='SEQCII_50pc__T_SRR7890936_50pc',
    genome_build='hg38',
    key_genes='/g/data/gx8/projects/diakumis/conda/envs/umccrise/lib/python3.7/site-packages/ngs_utils/reference_data/key_genes/umccr_cancer_genes.latest.tsv',
    af_global='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/work/SEQCII_50pc__T_SRR7890936_50pc/rmd/afs/af_tumor.txt',
    af_keygenes='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/work/SEQCII_50pc__T_SRR7890936_50pc/rmd/afs/af_tumor_keygenes.txt',
    somatic_snv='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/SEQCII_50pc__T_SRR7890936_50pc/small_variants/SEQCII_50pc__T_SRR7890936_50pc-somatic-PASS.vcf.gz',
    somatic_sv='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/SEQCII_50pc__T_SRR7890936_50pc/structural/SEQCII_50pc__T_SRR7890936_50pc-manta.tsv',
    somatic_sv_vcf='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/SEQCII_50pc__T_SRR7890936_50pc/structural/SEQCII_50pc__T_SRR7890936_50pc-manta.vcf.gz',
    purple_som_gene_cnv='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/work/SEQCII_50pc__T_SRR7890936_50pc/purple/SEQCII_50pc__T_SRR7890936_50pc.purple.cnv.gene.tsv',
    purple_som_cnv='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/work/SEQCII_50pc__T_SRR7890936_50pc/purple/SEQCII_50pc__T_SRR7890936_50pc.purple.cnv.somatic.tsv',
    purple_germ_cnv='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/work/SEQCII_50pc__T_SRR7890936_50pc/purple/SEQCII_50pc__T_SRR7890936_50pc.purple.cnv.germline.tsv',
    purple_purity='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/work/SEQCII_50pc__T_SRR7890936_50pc/purple/SEQCII_50pc__T_SRR7890936_50pc.purple.purity.tsv',
    purple_qc='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/work/SEQCII_50pc__T_SRR7890936_50pc/purple/SEQCII_50pc__T_SRR7890936_50pc.purple.qc',
    purple_som_snv_vcf='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/work/SEQCII_50pc__T_SRR7890936_50pc/purple/SEQCII_50pc__T_SRR7890936_50pc.purple.somatic.vcf.gz',
    oncoviral_present_viruses='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/work/SEQCII_50pc__T_SRR7890936_50pc/oncoviruses/present_viruses.txt',
    oncoviral_breakpoints_tsv='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/work/SEQCII_50pc__T_SRR7890936_50pc/oncoviruses/oncoviral_breakpoints.tsv',
    conda_list='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/work/conda_pkg_list.txt',
    result_outdir='/g/data/gx8/projects/diakumis/umccrise/test_umccrise/umccrised/SEQCII_50pc__T_SRR7890936_50pc/cancer_report_results'),
  local = list(
    genome_build='hg38',
    key_genes=glue::glue('{dd}/data/ref/umccr_cancer_genes.latest.tsv'),
    af_global=glue::glue('{dd}/data/umccrised/work/{batch_name}/cancer_report/afs/af_tumor.txt'),
    af_keygenes=glue::glue('{dd}/data/umccrised/work/{batch_name}/cancer_report/afs/af_tumor_keygenes.txt'),
    somatic_snv=glue::glue('{dd}/data/umccrised/{batch_name}/small_variants/{batch_name}-somatic-PASS.vcf.gz'),
    somatic_sv=glue::glue('{dd}/data/umccrised/{batch_name}/structural/{batch_name}-manta.tsv'),
    somatic_sv_vcf=glue::glue('{dd}/data/umccrised/{batch_name}/structural/{batch_name}-manta.vcf.gz'),
    purple_som_gene_cnv=glue::glue('{dd}/data/umccrised/work/{batch_name}/purple/{batch_name}.purple.cnv.gene.tsv'),
    purple_som_cnv=glue::glue('{dd}/data/umccrised/work/{batch_name}/purple/{batch_name}.purple.cnv.somatic.tsv'),
    purple_germ_cnv=glue::glue('{dd}/data/umccrised/work/{batch_name}/purple/{batch_name}.purple.cnv.germline.tsv'),
    purple_purity=glue::glue('{dd}/data/umccrised/work/{batch_name}/purple/{batch_name}.purple.purity.tsv'),
    purple_qc=glue::glue('{dd}/data/umccrised/work/{batch_name}/purple/{batch_name}.purple.qc'),
    purple_som_snv_vcf=glue::glue('{dd}/data/umccrised/work/{batch_name}/purple/{batch_name}.purple.somatic.vcf.gz'),
    oncoviral_present_viruses=glue::glue('{dd}/data/umccrised/work/{batch_name}/oncoviruses/present_viruses.txt'),
    oncoviral_breakpoints_tsv=glue::glue('{dd}/data/umccrised/work/{batch_name}/oncoviruses/oncoviral_breakpoints.tsv'),
    conda_list=glue::glue('{dd}/data/umccrised/work/conda_pkg_list.txt'),
    result_outdir=glue::glue('{dd}/data/umccrised/{batch_name}/cancer_report_results')
  )
)

params <- params_tmp[["local"]]
params$batch_name <- batch_name
params$tumor_name <- tumor_name

render_me <- function() {
  rmarkdown::render(
    "cancer_report.Rmd",
    params = params)
}
render_me()
```

```{r load_pkgs}
# Bioconductor
library(BSgenome)
library(MutationalPatterns)
ref_genome <- paste0("BSgenome.Hsapiens.UCSC.", params$genome_build)
library(ref_genome, character.only = TRUE)
tx_ref_genome <- paste0("TxDb.Hsapiens.UCSC.", params$genome_build, ".knownGene")
library(tx_ref_genome, character.only = TRUE)
# CRAN
library(devtools)
library(details)
library(DT)
library(dplyr)
library(glue)
library(gt)
library(ggplot2)
library(htmltools)
library(jsonlite)
library(knitr)
library(kableExtra)
library(patchwork)
library(purrr)
library(readr)
library(rmarkdown)
library(stringr)
library(tidyr)
# umccr
library(gpgr)
```

```{r funcs}
img_dir <- file.path("img") # created when copying purple files to tmp dir
result_outdir <- params$result_outdir # directory to write tables to
blank_lines <- function(n = 10){cat(rep("&nbsp;  ",n), sep="\n")}

# get js indices when using in DT options
col_ind_js <- function(dt, col) {
  ntab <- setNames((1:ncol(dt) - 1), names(dt))
  stopifnot(all(col %in% names(dt)))
  sort(unname(ntab[col]))
}

# used in SV table Details
sv_detail_table <- function(tab) {
  # for checking if the columns in the dt are all described
  check_cols_described <- function(cols, descr) {
    stopifnot(is.character(cols), is.atomic(cols))
    stopifnot(is.character(descr), is.atomic(descr))
    for (col in cols) {
      if (!col %in% descr) {
        warning(glue::glue("{col} is not described!"))
      }
    }
  }

  check_cols_described(colnames(tab), sv_col_descr$Column)
  list(Column = colnames(tab)) %>%
    tibble::as_tibble() %>%
    dplyr::left_join(sv_col_descr, by = "Column")
}
```


```{r hrd_prep, results='hide'}
#---- Homologous Recombination Deficiency ----#
chord_sv_df <- gpgr::chord_mantavcf2df(params$somatic_sv_vcf)
chord_res <- gpgr::chord_run(
  vcf.snv = params$somatic_snv,
  df.sv = chord_sv_df,
  sv.caller = "manta",
  sample.name = params$tumor_name,
  ref.genome = params$genome_build,
  verbose = TRUE)

hrdetect_res <- NULL
# HRDetect doesn't handle empty files
if (gpgr::vcf_is_empty(params$somatic_sv_vcf) |
    gpgr::vcf_is_empty(params$somatic_snv)) {
  hrdetect_res <- tibble::tribble(
    ~sample, ~Probability, ~intercept, ~del.mh.prop, ~SNV3, ~SV3, ~SV5, ~hrd, ~SNV8,
    params$tumor_name, NA, NA, NA, NA, NA, NA, NA, NA
  )
} else {
  hrdetect_res <- gpgr::hrdetect_run(
    nm = params$tumor_name,
    snvindel_vcf = params$somatic_snv,
    sv_vcf = params$somatic_sv_vcf,
    cnv_tsv = params$purple_som_cnv,
    genome = params$genome_build,
    snvoutdir = tempdir())
}
```

```{r mutational_sigs_prep}
#---- Mutational Signatures ----#
somatic_snv <- MutationalPatterns::read_vcfs_as_granges(
  vcf_files = params$somatic_snv,
  sample_names = params$tumor_name,
  genome = ref_genome,
  group = "auto+sex")

mut_mat <- MutationalPatterns::mut_matrix(vcf_list = somatic_snv, ref_genome = ref_genome)

###--- OLD Sigs ---###
# Get Sanger sigs from "http://cancer.sanger.ac.uk/cancergenome/assets/signatures_probabilities.txt"
sig_probs1 <- file.path("misc/sig/v2_mar2015/signatures_probabilities.txt")
# better be explicit - the sig_probs file has 7 extra empty columns
ctypes <- paste0(c("ccc", paste0(rep("d", 30), collapse = ""), "ccccccc"), collapse = "")
cnames <- c("SubstType", "Trinucleotide", "SomMutType", paste0("Sig", 1:30), paste0("foo", 1:7))
cancer_signatures1 <-
  readr::read_tsv(sig_probs1, col_names = cnames, col_types = ctypes, skip = 1) %>%
  dplyr::arrange(SubstType)

# sanity check - need to be in same order
stopifnot(all(cancer_signatures1$SomMutType == rownames(mut_mat)))
# select only 30 sig columns, 96 mut types
cancer_signatures1 <- cancer_signatures1 %>%
  dplyr::select(4:33) %>%
  as.matrix()

###--- NEW Sigs ---###
sig_probs2 <- file.path("misc/sig/v3_may2019/sigProfiler_SBS_signatures_2019_05_22.csv")
cancer_signatures2 <-
  readr::read_csv(sig_probs2, col_types = cols(.default = "d", Type = "c", SubType = "c")) %>%
  dplyr::mutate(SomMutType = paste0(substr(SubType, 1, 1),
                                    "[", Type, "]",
                                    substr(SubType, 3, 3))) %>%
  dplyr::select(SomMutType, SBS1:SBS85)

# sanity check - need to be in same order
stopifnot(all(cancer_signatures2$SomMutType == rownames(mut_mat)))
# select only 67 sig columns, 96 mut types
cancer_signatures2 <- cancer_signatures2 %>%
  dplyr::select(-1) %>%
  as.matrix()

# Fit mutation matrix to cancer signatures
fit_res1 <- MutationalPatterns::fit_to_signatures(mut_matrix = mut_mat, signatures = cancer_signatures1)
fit_res2 <- MutationalPatterns::fit_to_signatures(mut_matrix = mut_mat, signatures = cancer_signatures2)

# Convert to tibbles for more bullet-proof subsetting
fit_res1 <- tibble::tibble(
  sig = rownames(fit_res1$contribution), # matrix rownames
  contr = unname(fit_res1$contribution[, 1])) # matrix single column -> vector

fit_res2 <- tibble::tibble(
  sig = rownames(fit_res2$contribution),
  contr = unname(fit_res2$contribution[, 1]))

# Select signatures with some contribution
fit_res_contr1 <- dplyr::filter(fit_res1, contr > 0)
fit_res_contr2 <- dplyr::filter(fit_res2, contr > 0)

# If there are no signatures, return a dummy tibble
if (nrow(fit_res_contr1) == 0) {
  fit_res_contr1 <- tibble::tibble(
    sig = "NOSIGS",
    contr = 0
  )
}

if (nrow(fit_res_contr2) == 0) {
  fit_res_contr2 <- tibble::tibble(
    sig = "NOSIGS",
    contr = 0
  )
}

mut_sig_contr1 <-
  fit_res_contr1 %>%
  dplyr::mutate(contr = round(contr, 0),
                Freq = round(contr / sum(contr), 2),
                Rank = as.integer(base::rank(-contr))) %>%
  dplyr::select(Rank, Signature = sig, Contribution = contr, Freq) %>%
  dplyr::arrange(Rank)
mut_sig_contr2 <-
  fit_res_contr2 %>%
  dplyr::mutate(contr = round(contr, 0),
                Freq = round(contr / sum(contr), 2),
                Rank = as.integer(base::rank(-contr))) %>%
  dplyr::select(Rank, Signature = sig, Contribution = contr, Freq) %>%
  dplyr::arrange(Rank)

gpgr::write_tsvjsongz(mut_sig_contr1, "sigs/mutsig1", result_outdir)
gpgr::write_tsvjsongz(mut_sig_contr2, "sigs/mutsig2", result_outdir)
```

```{r sv_prep}
#---- Structural Variants ----#
sv_path <- params$somatic_sv
sv <- NULL
sv_unmelted <- NULL
sv_all <- NULL
no_sv_found <- TRUE
sv_tsv_descr <- NULL
sv_col_descr <- NULL
if (sv_path == "NA") {
  sv_unmelted <- tibble(WARNING = "Structural variants were not called")
  sv_all <- tibble(WARNING = "Structural variants were not called")
} else {
  no_sv_found <- gpgr::tsv_is_empty(sv_path)
  if (no_sv_found) {
    sv_unmelted <- tibble(WARNING = "THERE WERE 0 SVs PRIORITISED!")
    sv_all <- tibble(WARNING = "THERE WERE 0 SVs PRIORITISED!")
    sv_tsv_descr <- tibble(Column = "THERE WERE 0 SVs PRIORITISED!")
    sv_col_descr <- tibble(Column = "THERE WERE 0 SVs PRIORITISED!")
  } else {
    sv <- gpgr::process_sv(sv_path)
    sv_unmelted <- sv$unmelted
    sv_all <- sv$melted
    sv_tsv_descr <- sv$tsv_descr
    sv_col_descr <- sv$col_descr
  }
}

gpgr::write_tsvjsongz(sv_unmelted, "sv/01_sv_unmelted", result_outdir)
gpgr::write_tsvjsongz(sv_all, "sv/02_sv_melted", result_outdir)
```

```{r purple_prep}
#---- PURPLE QC Table ----#
purple_qc_summary <- dplyr::bind_rows(
  gpgr::purple_qc_read(params$purple_qc)[["summary"]],
  gpgr::purple_purity_read(params$purple_purity)[["summary"]]
)

#---- PURPLE CNV Tables ----#
purple_cnv_som <- gpgr::purple_cnv_som_process(params$purple_som_cnv)
purple_cnv_som_gene <- gpgr::purple_cnv_som_gene_process(params$purple_som_gene_cnv, params$key_genes)
purple_cnv_germ <- gpgr::purple_cnv_germ_process(params$purple_germ_cnv)

gpgr::write_tsvjsongz(purple_cnv_som$tab, "purple/purple_cnv_som", result_outdir)
gpgr::write_tsvjsongz(purple_cnv_som_gene$tab, "purple/purple_cnv_som_gene", result_outdir)
gpgr::write_tsvjsongz(purple_cnv_germ$tab, "purple/purple_cnv_germ", result_outdir)
```

```{r, results='asis'}
blank_lines(2)
```

## Summary

<details>
<summary>Description</summary>

PURPLE outputs a QC status along with a summary
for the inferred purity and ploidy of the somatic sample.
A failed QC status can be attributed to several factors.

__QC Status__

The QC Status field reflects how we have determined the purity of the sample:

* `NORMAL` - PURPLE fit the purity using COBALT and AMBER output.
* `HIGHLY_DIPLOID` - The fitted purity solution is highly diploid (> 95%)
  with a large range of potential solutions, but somatic variants are unable to
  help either because they were not supplied or because their implied purity was too low.
* `SOMATIC` - Somatic variants have improved the otherwise highly diploid solution.
* `NO_TUMOR` - PURPLE failed to find any aneuploidy and somatic variants were
  supplied but there were fewer than 300 with observed VAF > 0.1.

__QC Failure__

There are several reasons PURPLE may classify a sample as failed:

* `FAIL_SEGMENT`: more than 220 copy number segments __not__ supported at either end
  by SV breakpoints. Indicates samples with extreme GC bias, with differences
  in depth of >= 10x between high and low GC regions. GC normalisation is
  unreliable when the corrections are so extreme so it is recommended to fail
  the sample (concerns with miscalled deletions or amplifications or have
  poor sensitivity in high GC regions.
* `NO_TUMOR`: no aneuploidy found and the number of somatic SNVs found is
  less than 1,000
* `MIN_PURITY`: fitted purity < 20%
* `FAIL_DELETED_GENES`: more than 280 deleted genes. This QC step was added
  after observing that in a handful of samples with high MB scale positive GC
  bias we sometimes systematically underestimate the copy number in high GC
  regions. This can lead us to incorrectly infer homozygous loss of entire
  chromosomes, particularly on chromosomes 17 and 19.
* `FAIL_GENDER`: if the AMBER and COBALT inferred genders are inconsistent
  then the COBALT one is used but the sample is failed.

</details>


```{r summary_tables}
summarise_sigs <- function(mut_sig_contr) {
  mut_sig_contr %>%
    head(2) %>%
    dplyr::mutate(string = paste0(Signature, " (", Contribution, " - ", Freq, ")")) %>%
    dplyr::pull(string) %>%
    paste(collapse = ", ")
}

hrd_summary <- list(chord = chord_res[["prediction"]][, "p_hrd", drop = TRUE],
                    hrdetect = hrdetect_res[, "Probability", drop = TRUE])

cnv_summary <-
  list(som = purple_cnv_som$tab, germ = purple_cnv_germ$tab) %>%
  purrr::map(~ dplyr::summarise(
    .x,
    Min = round(min(CN), 2),
    Max = round(max(CN), 2),
    N = dplyr::n(), .groups = "drop_last"))

if (no_sv_found) {
  sv_summary <- tibble(unmelted = c(0), melted = c(0))
} else {
  sv_summary <-
    list(unmelted = sv_unmelted %>% dplyr::select(Type),
         melted = sv_all %>% dplyr::select(Type)) %>%
    purrr::map(function(x) addmargins(table(x, useNA = 'ifany'))) %>%
    do.call("rbind", .) %>%
    as.data.frame() %>%
    tibble::rownames_to_column(var = "group") %>%
    tidyr::pivot_longer(-group) %>%
    dplyr::mutate(value = paste0(name, ": ", value)) %>%
    tidyr::pivot_wider(names_from = group, values_from = value) %>%
    dplyr::select(unmelted, melted)
}

qc_summary_all <- dplyr::tribble(
  ~n, ~variable, ~value, ~details,
  5, "MutSigs (Old)", glue::glue("{summarise_sigs(mut_sig_contr1)}"),
  "v2 COSMIC (https://cancer.sanger.ac.uk/cosmic/signatures_v2.tt)",
  5, "MutSigs (New)", glue::glue("{summarise_sigs(mut_sig_contr2)}"),
  "v3 COSMIC (https://cancer.sanger.ac.uk/cosmic/signatures)",
  6, "HRD", glue::glue("CHORD: {hrd_summary$chord}; HRDetect: {hrd_summary$hrdetect}"), "",
  16, "CNVs (Somatic)", glue::glue("Min: {cnv_summary[['som']]$Min}; ",
                                   "Max: {cnv_summary[['som']]$Max}; ",
                                   "N: {cnv_summary[['som']]$N} "), "",
  16, "CNVs (Germline)", glue::glue("Min: {cnv_summary[['germ']]$Min}; ",
                                   "Max: {cnv_summary[['germ']]$Max}; ",
                                   "N: {cnv_summary[['germ']]$N} "), "",
  17, "SVs (unmelted)", paste(rev(sv_summary$unmelted), collapse = ", "), "",
  18, "SVs (melted)", paste(rev(sv_summary$melted), collapse = ", "), "",
  19, "Genome", params$genome_build, ""
) %>%
  dplyr::bind_rows(
    purple_qc_summary
  ) %>%
  dplyr::arrange(n) %>%
  dplyr::select(-n)

qc_summary_all %>%
  gt::gt(rowname_col = "variable") %>%
  gt::tab_style(
    style = list(
      cell_fill(color = "#FF4646"),
      cell_text(weight = "bold")
    ),
    locations = cells_body(
      columns = vars(value),
      rows = grepl("FAIL", value) & variable == "QC_Status")) %>%
  gt::tab_style(
    style = list(
      cell_fill(color = "#FF4646"),
      cell_text(weight = "bold")
    ),
    locations = cells_body(
      columns = vars(value),
      rows = grepl("FALSE", value) & variable %in% c("Segment_Pass", "Gender_Pass", "DelGenes_Pass"))) %>%
  gt::tab_style(
    style = list(
      cell_text(weight = "bold")
    ),
    locations = cells_stub(rows = TRUE)
  ) %>%
  gt::cols_align("left") %>%
  gt::opt_row_striping() %>%
  gt::tab_options(table.align = "left")
```

```{r, results='asis'}
blank_lines(1)
```

## Somatic Mutation Profiles

```{r, results='asis'}
blank_lines(1)
```

### Allelic Frequencies {#allelic-freqs}
Summarised below are the allele frequencies (AFs) for somatic variants detected
genome-wide (__Global__) vs. within the coding sequence of ~1,100 UMCCR
cancer genes (__Key Genes CDS__).
AFs range from 0 to 1, or 0%-100% (we filter out all novel variants with AF < 10%).

<details>
<summary>Details</summary>

Variants are typically called in bcbio by 3 different callers, with calls supported by at least 2 of them used ("ensemble" approach).
In some cases only a single caller is used due to technical reasons (e.g. highly mutated FFPE sample).

The following post-processing steps occur:

1. `somatic_vcf_annotate`: [annotate VCF](https://github.com/umccr/vcf_stuff/blob/master/vcf_stuff/filtering/annotate_somatic_vcf.smk)
    against databases of known hotspots, germline variants, low mappability regions, UMCCR panel of normals
2. `somatic_vcf_filter`: [filter VCF](https://github.com/umccr/vcf_stuff/blob/master/scripts/filter_somatic_vcf)
    to remove germline variants and artefacts, but keep known hotspots
3. As preparation for the allelic frequencies plots:
    * `subset_to_giab`: keep variants in 'high confidence' regions as determined by the [Genome in a Bottle consortium](http://jimb.stanford.edu/giab/)
    * keep only variants with AF above 10%
4. Allele frequencies for global and keygenes:
    * `afs`: grab only the `INFO/TUMOR_AF` field and output to final txt file
    * `afs_keygenes`: grab the `CHROM`, `POS`, `ID`, `REF`, `ALT` and `INFO/TUMOR_AF`
      for variants in the UMCCR cancer gene BED file, and output to final txt file
</details>


```{r allele_freq_stats}
af_summary_results <- gpgr::af_summary(params$af_global, params$af_keygenes)
af_summary_results$af_stats_gt
```

```{r, results='asis'}
blank_lines(1)
```

```{r allele_freq_plot, fig.width=10, fig.height=3}
af_summary_results$af_plot
```

```{r, results='asis'}
blank_lines(1)
```

### Mutational Signatures {.tabset .tabset-fade #signatures}

Deciphering the mutational signature of a tumor sample can provide insight into the mutational
processes involved in carcinogenesis and help in cancer treatment and prevention.
The [MutationalPatterns](http://bioconductor.org/packages/release/bioc/html/MutationalPatterns.html)
R package is used to generate a mutation signature for the sample. We use the final filtered somatic
calls as input.

#### Context signature

```{r plot_96_prof, fig.height=3}
MutationalPatterns::plot_96_profile(mut_mat, condensed = TRUE)
```

#### Point mutation spectrum

<details>
<summary>Description</summary>

We can count the mutation type occurrences for the input VCF.
For `C>T` mutations, a distinction is made between `C>T` at CpG sites
and other sites, as deamination of methylated cytosine at CpG sites is a common mutational
process. This is the reason the reference genome is needed.

A mutation spectrum shows the relative contribution of each mutation type in the base
substitution catalogs. We can plot the mean relative contribution of
each of the 6 base substitution types over all samples. Error bars indicate standard deviation
over all samples. The total number of mutations is indicated. We can also distinguish
between `C>T` at CpG sites and other sites.

</details>

```{r plot_mut_type_occurrences, fig.height=3}
type_occurrences <- MutationalPatterns::mut_type_occurrences(vcf_list = somatic_snv, ref_genome = ref_genome)
MutationalPatterns::plot_spectrum(type_occurrences, CT = TRUE)
```

#### Transcriptional strand bias

<details>
<summary>Description</summary>

We can determine if a gene mutation is on the transcribed or non-transcribed
strand, which can be used to evaluate the involvement of transcription-coupled
repair. By convention base substitutions are regarded as C>X or T>X, so we try
to determine whether the C or T base is on the same strand as the gene
definition. Base substitutions on the same strand as the gene definition are
considered 'untranscribed', and on the opposite strand 'transcribed', since the
gene definitions report the coding or sense strand, which is untranscribed. No
strand information is reported for base substitutions that overlap with more
than one gene on different strands.

</details>

```{r tran_strand_bias, fig.height=3}
# Get known genes table from UCSC
if (params$genome_build == 'hg19') {
  genes_list <- GenomicFeatures::genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
} else {
  genes_list <- GenomicFeatures::genes(TxDb.Hsapiens.UCSC.hg38.knownGene)
}

# Mutation count matrix with strand info (4*6*4=96 -> 96*2=192)
mut_mat_s <- MutationalPatterns::mut_matrix_stranded(somatic_snv,
                                                     ref_genome = ref_genome,
                                                     ranges = genes_list,
                                                     mode = "transcription")

# Mutation count per type and strand
strand_counts <- MutationalPatterns::strand_occurrences(mut_mat_s, by = "all")
# Poisson test for strand asymmetry significance testing
strand_bias <- MutationalPatterns::strand_bias_test(strand_counts)

# mutation spectrum with strand distinction
MutationalPatterns::plot_strand(strand_counts, mode = "relative")
# effect size of strand bias
MutationalPatterns::plot_strand_bias(strand_bias)
```

#### Replicative strand bias

<details>
<summary>Description</summary>

The involvement of replication-associated mechanisms can be evaluated by
testing for a mutational bias between the leading and lagging strand.
The replication strand is dependent on the locations of replication
origins from which DNA replication is fired.
However, replication timing is dynamic and cell-type specific,
which makes replication strand determination less straightforward than
transcriptional strand bias analysis.
Replication timing profiles can be generated with
Repli-Seq experiments. Once the replication direction is defined,
a strand asymmetry analysis can be performed similarly as the transcription
strand bias analysis.

</details>

```{r rep_strand_bias, fig.height=3}
repli_file <- system.file("extdata/ReplicationDirectionRegions.bed",
                          package = "MutationalPatterns")
# start/stop contain scientific notation, so need to be doubles
repli_strand <-
  readr::read_tsv(repli_file, col_names = TRUE, col_types = "cddcc") %>%
  dplyr::mutate_if(is.character, as.factor)
repli_strand_granges <- GenomicRanges::GRanges(
  seqnames = repli_strand$Chr,
  ranges = IRanges::IRanges(start = repli_strand$Start + 1,
                   end = repli_strand$Stop),
  strand_info = repli_strand$Class)

GenomeInfoDb::seqlevelsStyle(repli_strand_granges) <- GenomeInfoDb::seqlevelsStyle(base::get(ref_genome))

mut_mat_s_rep <-
  MutationalPatterns::mut_matrix_stranded(
    vcf_list = somatic_snv,
    ref_genome = ref_genome,
    ranges = repli_strand_granges,
    mode = "replication")
# Mutation count per type and strand
strand_counts_rep <- MutationalPatterns::strand_occurrences(mut_mat_s_rep, by = "all")
# Poisson test for strand asymmetry significance testing
strand_bias_rep <- MutationalPatterns::strand_bias_test(strand_counts_rep)

MutationalPatterns::plot_strand(strand_counts_rep, mode = "relative")
MutationalPatterns::plot_strand_bias(strand_bias_rep)
```

```{r, results='asis'}
blank_lines(1)
```

### Signature Contribution {.tabset .tabset-fade #contributions}

<details>
<summary>Description</summary>

The contribution of any set of signatures to the mutational profile of a
sample can be quantified. This unique feature is specifically useful
for mutational signature analyses of small cohorts or individual samples,
but also to relate own findings to known signatures and published findings.
The `fit_to_signatures` function finds the optimal linear combination of
mutational signatures that most closely reconstructs
the mutation matrix by solving a non-negative least-squares constraints problem.

Shown are signatures with positive Contribution values, along with summarised descriptions
and reference signature plots from <https://cancer.sanger.ac.uk/cosmic/signatures>.

</details>

```{r, results='asis'}
blank_lines(1)
```

#### OLD

```{r mutational_signature_contribution}
sig_table <-
  readr::read_tsv(file = "misc/sig/v2_mar2015/signatures_description.tsv", col_types = "cc") %>%
  dplyr::mutate(Plot = paste0("![](misc/sig/v2_mar2015/img/sig-", signature, ".png)"),
                signature = paste0("Sig", signature)) %>%
  dplyr::select(Signature = signature, Description = description, Plot)

mut_sig_contr1 %>%
  dplyr::left_join(sig_table, by = "Signature") %>%
  knitr::kable() %>%
  kableExtra::kable_styling(c("hover", "striped"), font_size = 12) %>%
  kableExtra::scroll_box(height = "400px")
```

```{r, results='asis'}
blank_lines(1)
```


#### NEW

```{r mutational_signature_contribution2}
sig_table2 <-
  readr::read_tsv(file = "misc/sig/v3_may2019/signatures_description.tsv", col_types = "cc") %>%
  dplyr::mutate(Plot = paste0("![](misc/sig/v3_may2019/img/sbs", signature, ".png)"),
                signature = paste0("SBS", signature)) %>%
  dplyr::select(Signature = signature, Description = description, Plot)

possible_seq_artefacts <- c("SBS27", "SBS43", "SBS45", "SBS46", "SBS47", "SBS48", "SBS49", "SBS50", "SBS51", "SBS52",
                            "SBS53", "SBS54", "SBS55", "SBS56", "SBS57", "SBS58", "SBS59", "SBS60")

mut_sig_contr2 %>%
  dplyr::left_join(sig_table2, by = "Signature") %>%
  dplyr::mutate(Signature = ifelse(Signature %in% possible_seq_artefacts, paste0(Signature, " (SA)"), Signature)) %>%
  knitr::kable() %>%
  kableExtra::kable_styling(c("hover", "striped"), font_size = 12) %>%
  kableExtra::scroll_box(height = "400px")
```

```{r, results='asis'}
blank_lines(1)
```

### Rainfall Plot {#rainfall}
Rainfall plots show the distribution of mutations along the genome, with mutation types
indicated with different colors. The y-axis corresponds to the distance of a mutation from the
previous mutation, and is log10 transformed. Drop-downs from the plots indicate clusters or
"hotspots" of mutations.

```{r mutationalpatterns_rainfall, out.width="90%", fig.width=14}
# When there is only 1 or lower number of variants on a chromosome,
# MutationalPatterns::plot_rainfall will crash with an error. So need to check if it will work beforehand.
chromosomes = GenomeInfoDb::seqnames(base::get(ref_genome))[1:22]
vcf = somatic_snv[[1]]
chr_subset = vcf[GenomeInfoDb::seqnames(vcf) == chromosomes[1]]
n = length(chr_subset)
will_work = FALSE
for (i in 1:length(chromosomes)) {
  chr_subset = vcf[GenomeInfoDb::seqnames(vcf) == chromosomes[i]]
  n = length(chr_subset)
  if (n >= 2) {
    will_work = TRUE
  }
}
if (will_work) {
  MutationalPatterns::plot_rainfall(somatic_snv[[1]],
                                    chromosomes = GenomeInfoDb::seqnames(base::get(ref_genome))[1:22],
                                    cex = 1.2, ylim = 1e+09)
}
```

### Kataegis
Somatic variants of type C>T and C>G in a TpCpN context are annotated as
showing Kataegis (Nik Zainal et al., 2012) if there are three or more
mutations of that type, strand and context localised within a
region with an average inter-mutation distance of <= 1kb.
The annotation describes the strand it was found on
along with an ID that is shared by grouped variants.

```{r purple_kataegis_table}
kat <- gpgr::purple_kataegis(params$purple_som_snv_vcf)
if (nrow(kat) > 0) {
  knitr::kable(kat, escape = FALSE, caption = "Kataegis Regions.",
               format = "html", format.args = list(big.mark = ",")) %>%
    kableExtra::kable_minimal("hover", full_width = FALSE, position = "left") %>%
    kableExtra::scroll_box(height = "150px")
} else {
  cat("No kataegis regions detected.")
}
```

```{r purple_plot_somaticrainfall, out.width="70%"}
knitr::include_graphics(file.path(img_dir, paste0(params$batch_name, '.somatic.rainfall.png')))
```

```{r, results='asis'}
blank_lines(1)
```

## HRD (Homologous Recombination Deficiency) Detection {#hrd}

<details>
<summary>Description</summary>

__CHORD__

For more details see <https://github.com/UMCUGenetics/CHORD>.

__Probabilities__

CHORD outputs the probability of:

* HRD: `p_hrd`
* BRCA1-type HRD: `p_BRCA1`
* BRCA2-type HRD: `p_BRCA2`

__HR status and type__

`hr_status` tells us if a sample is HR deficient (`p_hrd >= 0.5`) or proficient.
If a sample is HRD, then `hrd_type` will tell us if the sample has BRCA1-type HRD
or BRCA2-type HRD (=`max(p_BRCA1, p_BRCA2)`).

__Pre-requisites for accurate HRD prediction__

* CHORD requires `>=50` indels to accurately determine whether a sample is HRD

If this criterion is not met, `hr_status` will be `cannot_be_determined` and `remarks_hr_status` will be
`<50 indels`.

* CHORD cannot be applied to MSI samples

If an MSI sample is detected, `hr_status` will be `cannot_be_determined` and `remarks_hr_status` will be
`Has MSI (>14000 indel.rep)`.

* If a sample is HRD, CHORD requires `>=30` SVs to accurately determine HRD subtype

If this criterion is not met, `hrd_type` will be `cannot_be_determined`, and `remarks_hr_status` will be `<30 SVs`.

The user may of course ignore these remarks and proceed with using the raw probabilities output by CHORD
(`p_hrd` and/or `p_BRCA1/pBRCA2`).

__HRDetect__

See https://github.com/Nik-Zainal-Group/signature.tools.lib for the R package
that includes HRDetect.

</details>

```{r chord_hrdetect_res}
hrd_results <- gpgr::hrd_results_tabs(hrdetect_res, chord_res)
gpgr::write_tsvjsongz(hrd_results$hrd_results_tab, "hrd/chord_hrdetect", result_outdir)
hrd_results$hrd_results_gt
```

```{r, results='asis'}
blank_lines(1)
```

## Circos Plots {.tabset .tabset-fade #circos}
Circos plots are generated by PURPLE.
The first BAF plot is based on PURPLE data and configuration files.

### BAF, Total/Minor CN, SVs

<details>
<summary>Description</summary>

* __Track1__: Chromosomes. Darker shaded areas: gaps in reference genome
  (centromeres, heterochromatin & missing short arms)
* __Track2__: <span style="color:#8A2BE2">Beta Allele Frequency</span>.
  Given that the BAF points correspond to allele frequencies of heterozygous SNPs that are common in germline samples,
  there shouldn't be any in chromosome Y (and chromosome X when male).
* __Track3__: __Total__ copy number changes adjusted for tumor purity,
  including focal and chromosomal somatic events.
  <span style="color:red">Red</span> = Loss; <span style="color:#32CD32">Green</span> = Gain.
  Scaled from 0 (complete loss) to 6 (high level gains).
  If > 6, shown as 6 with a green dot on the outermost green gridline.
* __Track4__: __Minor__ allele copy numbers. Range from 0 to 3.
  Expected normal minor allele copy number is 1, and anything below 1 is shown
  as a loss (<span style="color:#EE7600">Orange</span>), representing an LOH event.
  Minor allele copy numbers above 1 (<span style="color:#7EC0EE">Blue</span>) indicate gains
  of both A and B alleles.
* __Track5__ (Inner circle): Observed structural variants within or between the chromosomes.
    * <span style="color:#7EC0EE">Blue</span> = Translocations
    * <span style="color:red">Red</span> = Deletions
    * <span style="color:#e6e600">Yellow</span> = Insertions
    * <span style="color:#32CD32">Green</span> = Tandem duplications
    * <span style="color:#000000">Black</span> = Inversions

</details>

```{r circos_baf_plot, out.width="80%"}
knitr::include_graphics(file.path(img_dir, paste0(params$batch_name, '.circos_baf.png')))
```


### SNVs/Indels, Total/Minor CN, SVs

<details>
<summary>Description</summary>

* __Track1__: Chromosomes. Darker shaded areas: gaps in reference genome
  (centromeres, heterochromatin & missing short arms)
* __Track2__: Somatic variants (incl. exon, intron and intergenic regions).
    * outer ring: SNP allele frequencies, corrected for tumor purity and scaled from 0 to 100%.
      Each dot represents a single somatic variant, coloured according to the
      type of base change (e.g. C>T/G>A in red).
    * inner ring: short insertion (yellow) and deletion (red) locations.
* __Track3__: Observed __total__ copy number changes adjusted for tumor purity,
  including focal and chromosomal somatic events.
  <span style="color:red">Red</span> = Loss; <span style="color:#32CD32">Green</span> = Gain.
  Scaled from 0 (complete loss) to 6 (high level gains).
  If > 6, shown as 6 with a green dot on the outermost green gridline.
* __Track4__: Observed __minor__ allele copy numbers. Range from 0 to 3.
  Expected normal minor allele copy number is 1, and anything below 1 is shown
  as a loss (<span style="color:#EE7600">Orange</span>), representing an LOH event.
  Minor allele copy numbers above 1 (<span style="color:#7EC0EE">Blue</span>) indicate gains
  of both A and B alleles.
* __Track5__ (Inner circle): Observed structural variants within or between the chromosomes.
    * <span style="color:#7EC0EE">Blue</span> = Translocations
    * <span style="color:red">Red</span> = Deletions
    * <span style="color:#e6e600">Yellow</span> = Insertions
    * <span style="color:#32CD32">Green</span> = Tandem duplications
    * <span style="color:#000000">Black</span> = Inversions

</details>

```{r circos_default1, out.width="80%"}
knitr::include_graphics(file.path(img_dir, paste0(params$batch_name, '.circos.png')))
```

### Allele Ratios, BAF

<details>
<summary>Description</summary>

* __Track1__: Chromosomes. Darker shaded areas: gaps in reference genome
  (centromeres, heterochromatin & missing short arms)
* __Track2__: <span style="color:blue">Tumor</span> and <span style="color:#32CD32">Normal</span> Allele Ratios
* __Track3__: <span style="color:#EE7600">Beta Allele Frequency</span>
  Given that the BAF points correspond to allele frequencies of heterozygous SNPs that are common in germline samples,
  there shouldn't be any in chromosome Y (and chromosome X when male).

</details>

```{r circos_default2, out.width="80%"}
knitr::include_graphics(file.path(img_dir, paste0(params$batch_name, '.input.png')))
```

## Structural Variants {#sv-summary}
Structural variants are inferred with [Manta](https://github.com/illumina/manta),
adjusted using [PURPLE](https://github.com/hartwigmedical/hmftools/tree/master/purity-ploidy-estimator#7-structural-variant-recovery-and-single-breakend-filtering),
and prioritised using [simple_sv_annotation](https://github.com/AstraZeneca-NGS/simple_sv_annotation).
Allele frequencies, copy number changes and ploidy are purity-adjusted.

<details>
<summary>Details</summary>

The input file corresponds to `umccrised/<batch>/structural/<batch>-manta.tsv`.

It's generated through the following steps:

__Step 1: Processing__

* __Input__: Manta structural variant calls from bcbio
(`final/<tumor-name>/<batch-name>-sv-prioritize-manta.vcf.gz` (or
`<batch-name>-manta.vcf.gz` if not prioritised))
* Remove following annotations from Manta VCF:
  'INFO/SIMPLE_ANN', 'INFO/SV_HIGHEST_TIER', 'FILTER/Intergenic', 'FILTER/MissingAnn', 'FILTER/REJECT'
* Prioritise variants with [simple_sv_annotation](https://github.com/vladsaveliev/simple_sv_annotation)`
* __Output__: `work/{batch}/structural/prioritize/{batch}-manta.vcf.gz`
* Keep PASS variants
* If more than 100,000 variants, keep only variants where `INFO/SV_TOP_TIER <= 3`
* __Output__: `work/{batch}/structural/keep_pass/{batch}-manta.vcf`
* Deal with chromosome capitalisation occurring from SnpEff
* Run BreakPointInspector (BPI) if it was disabled in bcbio
* __Output__: `work/{batch}/structural/maybe_bpi/{batch}-manta.vcf`

__Step 2: Filtering__

* Keep PASS variants (since BPI updates the FILTER column)
* For BND variants require paired reads support (PR) to be higher than split read support (SR)
* Keep all `INFO/SV_TOP_TIER <= 2` variants
* For `INFO/SV_TOP_TIER > 2` variants require split _or_ paired reads support of at least 5x
* For `INFO/SV_TOP_TIER > 2` variants with low allele frequency at any breakpoint (`BPI_AF[0 or 1] < 0.1`),
  require SR or PR support of at least 10x
* __Output__: `work/{batch}/structural/filt/{batch}-manta.vcf`

__Step 3: PURPLE and FFPE conditional__

* If the sample _is not_ FFPE:
  * Feed above filtered SVs to PURPLE, which outputs `purple.sv.vcf.gz` that contains rescued SVs
  * Prioritise variants (again)
  * Remove 'INFO/ANN' annotation
  * __Output__: `{batch}/structural/{batch}-manta.vcf.gz`
* If the sample _is_ FFPE:
  * Just copy `filtered` variants and don't do anything
    (i.e. we don't want the rescued SVs from PURPLE since they'll likely be heaps)
    (note that PURPLE will still get fed with the `filtered` SVs)
  * __Output__: `{batch}/structural/{batch}-manta.vcf.gz`

__Step 4: TSV final output__

* Input: `{batch}/structural/{batch}-manta.vcf.gz` VCF
* Output: `{batch}/structural/{batch}-manta.tsv` TSV


```{r manta_tsv_description}
sv_tsv_descr %>%
  dplyr::arrange(Column) %>%
  dplyr::mutate(Column = kableExtra::cell_spec(Column, bold = TRUE)) %>%
  knitr::kable(escape = FALSE, caption = "Description of Manta TSV columns") %>%
  kableExtra::kable_paper(c("hover", "striped"), full_width = FALSE, position = "left") %>%
  kableExtra::scroll_box(height = "250px")
```

__Prioritisation process__

1. Annotate with [SnpEff](http://snpeff.sourceforge.net/SnpEff_manual.html) based on Ensembl gene model
2. Subset annotations to [APPRIS principal transcripts](http://appris.bioinfo.cnio.es/#/)
3. Prioritize variants with [simple_sv_annotation](https://github.com/vladsaveliev/simple_sv_annotation)
   1(high)-2(moderate)-3(low)-4(no interest):
  * exon loss
     - on prioritisation gene list (1)
     - other (2)
  * gene_fusion
     - paired (hits two genes)
        - on list of known pairs (1) (curated by [HMF](https://resources.hartwigmedicalfoundation.nl))
        - one gene is a known promiscuous fusion gene (1) (curated by [HMF](https://resources.hartwigmedicalfoundation.nl))
        - on list of FusionCatcher known pairs (2)
        - other:
           - one or two genes on prioritisation gene list (2)
           - neither gene on prioritisation gene list (3)
     - unpaired (hits one gene)
         - on prioritisation gene list (2)
         - others (3)
  * upstream or downstream
     - on prioritisation gene list genes (2)  - e.g. one gene is got into
       control of another gene's promoter and get overexpressed (oncogene) or underexpressed (tsgene)
  * LoF or HIGH impact in a tumor suppressor
     - on prioritisation gene list (2)
     - other TS gene (3)
  * other (4)

* Use PURPLE copy number caller to infer more SV calls from copy number transitions (marked as 'PURPLE Inferred')

```{r effect_abbrev_tab}
gpgr::EFFECT_ABBREVIATIONS %>%
  tibble::enframe(value = "abbreviation") %>%
  dplyr::arrange(abbreviation) %>%
  dplyr::mutate(name = kableExtra::cell_spec(name, bold = TRUE)) %>%
  knitr::kable(escape = FALSE) %>%
  kableExtra::kable_paper(c("hover", "striped"), full_width = FALSE, position = "left") %>%
  kableExtra::scroll_box(height = "250px")
```

</details>

### Summary

```{r plot_bnd_sr_pr_tot_lines, fig.width=12, fig.height=6, out.width="90%"}
p1 <- gpgr:::plot_bnd_sr_pr_tot_lines(sv_unmelted)
p2 <- gpgr::plot_bnd_sr_pr_tot_hist(sv_unmelted) + ggplot2::theme(legend.position = "none")
(p1 | p2) + patchwork::plot_layout(widths = c(3, 2))
```

```{r sv_summary, eval=!no_sv_found}
d <-
  list(unmelted = sv_unmelted %>% dplyr::select(Tier = TierTop, Type),
       melted = sv_all %>% dplyr::select(Tier, Type)) %>%
  dplyr::bind_rows(.id = "group") %>%
  dplyr::mutate(group = factor(group, levels = c("unmelted", "melted"))) %>%
  dplyr::group_by(group) %>%
  dplyr::count(Tier, Type) %>%
  tidyr::pivot_wider(names_from = Type, values_from = n) %>%
  dplyr::ungroup() %>%
  dplyr::rowwise() %>%
  dplyr::mutate(tot = sum(dplyr::c_across(!(c(group, Tier))), na.rm = TRUE))

d %>%
  gt::gt(
    groupname_col = "group",
    rowname_col = "Tier") %>%
  gt::tab_header(
    title = "SV type counts by tier",
    subtitle = "(TopTier used for melted variants)"
  ) %>%
  gt::tab_stubhead(label = "Tier") %>%
  gt::summary_rows(groups = TRUE,
                   fns = list(tot = ~sum(., na.rm = TRUE)),
                   drop_trailing_zeros = TRUE) %>%
  gt::fmt_missing(columns = dplyr::everything()) %>%
  gt::tab_style(
    style = list(
      cell_text(weight = "bold")
    ),
    locations = cells_stub(rows = TRUE)
  ) %>%
  gt::cols_align("right") %>%
  gt::tab_options(table.align = "left")
```

```{r, results='asis'}
blank_lines(1)
```

### Unmelted Variants {#sv-unmelted}

```{r svtab_raw, eval=!no_sv_found}
svtab_raw <- sv_unmelted %>%
  dplyr::select(-annotation) %>%
  dplyr::arrange(TierTop, vcfnum)

svtab_rawdt <- svtab_raw %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 400, scrollX = TRUE, autoWidth = FALSE, keys = TRUE,
                  buttons = c('csv'), dom = 'Blfrtip'))
```

```{r svtab_raw_details, eval=!no_sv_found}
details::details(sv_detail_table(svtab_raw))
svtab_rawdt
```

```{r, results='asis'}
blank_lines(2)
```

### Translocations (BNDs) {.tabset .tabset-fade #translocations}

```{r svtab_BND, eval=!no_sv_found}
sv_BND <- sv_all %>%
  dplyr::filter(Type == "BND") %>%
  dplyr::mutate(nrow = dplyr::row_number(),
                nrow = sprintf(glue::glue("%0{nchar(nrow(.))}d"), nrow)) %>%
  dplyr::select(nrow, dplyr::everything())


sv_BND_main <- sv_BND %>%
  dplyr::select(nrow, vcfnum, Tier, Start, End,
                ID, BND_ID, BND_mate,
                Genes, Effect, Detail, SR_alt, PR_alt, Ploidy, AF_PURPLE)

sv_BND_other <- sv_BND %>%
  dplyr::select(nrow, AF_BPI, CNC, CN, SR_PR_ref, SScore, ntrx, Transcript)

sv_BND_purple <- sv_all %>%
  dplyr::filter(Type %in% c("PURPLE_inf")) %>%
  dplyr::select(Tier, Start, Effect, Detail, Ploidy, CN, CNC, ID)

gpgr::write_tsvjsongz(
  dplyr::bind_cols(sv_BND_main, sv_BND_other %>% dplyr::select(-nrow)),
  "sv/03_sv_BND_main", result_outdir)
gpgr::write_tsvjsongz(sv_BND_purple, "sv/04_sv_BND_purpleinf", result_outdir)

sv_BND_maindt <- sv_BND_main %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 400, scrollX = TRUE, autoWidth = FALSE, keys = TRUE,
                  buttons = c('csv'), dom = 'Blfrtip'))

sv_BND_otherdt <- sv_BND_other %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 400, scrollX = TRUE, keys = TRUE, autoWidth = FALSE,
                  buttons = c('csv'), dom = 'Blfrtip'))

sv_BND_purpledt <- sv_BND_purple %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                options = list(
                  scroller = TRUE, scrollY = 400, scrollX = TRUE, autoWidth = TRUE,
                  buttons = c('csv'),
                  dom = 'Blfrtip'))
```

#### Main Columns

```{r sv_BND_main, eval=!no_sv_found}
details::details(sv_detail_table(sv_BND_main))
sv_BND_maindt
```

```{r, results='asis'}
blank_lines(2)
```

#### Other Columns

```{r sv_BND_otherdt, eval=!no_sv_found}
details::details(sv_detail_table(sv_BND_other))
sv_BND_otherdt
```

#### PURPLE Inferred

```{r sv_BND_purple, eval=!no_sv_found}
details::details(sv_detail_table(sv_BND_purple))
sv_BND_purpledt
```

### DEL/DUP/INS {.tabset .tabset-fade #deldupins}

```{r svtab_noBND, eval=!no_sv_found}
sv_noBND <- sv_all %>%
  dplyr::filter(!Type %in% c("BND", "PURPLE_inf")) %>%
  dplyr::mutate(nrow = dplyr::row_number(),
                nrow = sprintf(glue::glue("%0{nchar(nrow(.))}d"), nrow),
                Genes_original = Genes,
                Transcript_original = Transcript) %>%
  dplyr::select(-ID)

max_genes <- 2
max_transcripts <- 2

sv_noBND_main <- sv_noBND %>%
  dplyr::mutate(Genes = ifelse(ngen > max_genes,
                               glue::glue("Many Genes ({ngen})"),
                               Genes),
                Transcript = ifelse(ntrx > max_transcripts,
                                    glue::glue("Many Transcripts ({ntrx})"),
                                    Transcript)) %>%
  dplyr::select(nrow, vcfnum, TierTop, Tier, Type, Start, End, Effect,
                Genes, Transcript, Detail, SR_alt, PR_alt, AF_PURPLE)

gpgr::write_tsvjsongz(sv_noBND_main, "sv/05_sv_noBND_main", result_outdir)

sv_noBND_other <- sv_noBND %>%
  dplyr::select(vcfnum, Ploidy, AF_BPI, CNC, CN, SR_PR_ref, SScore) %>%
  dplyr::distinct()

gpgr::write_tsvjsongz(sv_noBND_other, "sv/06_sv_noBND_other", result_outdir)

sv_noBND_manygenes <- sv_noBND %>%
  dplyr::filter(ngen > max_genes) %>%
  dplyr::select(nrow, vcfnum, Tier, Type, Start, End,
                Effect, ngen, Genes = Genes_original)

gpgr::write_tsvjsongz(sv_noBND_manygenes, "sv/07_sv_noBND_manygenes", result_outdir)

sv_noBND_manytx <- sv_noBND %>%
  dplyr::filter(ntrx > max_transcripts) %>%
  dplyr::select(nrow, vcfnum, Tier, Type, Start, End,
                Effect, ntrx, Transcript = Transcript_original)

gpgr::write_tsvjsongz(sv_noBND_manytx, "sv/08_sv_noBND_manytranscripts", result_outdir)

sv_noBND_maindt <- sv_noBND_main %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Buttons"),
                options = list(
                  pageLength = 20,
                  autoWidth = FALSE,
                  buttons = c('csv'), dom = 'Blfrtip'))

sv_noBND_otherdt <- sv_noBND_other %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Buttons"),
                options = list(
                  pageLength = 20,
                  autoWidth = FALSE,
                  buttons = c('csv'),
                  dom = 'Blfrtip'))

sv_noBND_manygenesdt <- sv_noBND_manygenes %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Buttons", "KeyTable"),
                options = list(
                  pageLength = 10,
                  autoWidth = FALSE, keys = TRUE,
                  buttons = c('csv'),
                  dom = 'Blfrtip'))

sv_noBND_manytxdt <- sv_noBND_manytx %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Buttons", "KeyTable"),
                options = list(
                  pageLength = 10,
                  autoWidth = FALSE, keys = TRUE,
                  buttons = c('csv'),
                  dom = 'Blfrtip'))
```

#### Main Columns

```{r sv_noBND_main, eval=!no_sv_found}
details::details(sv_detail_table(sv_noBND_main))
sv_noBND_maindt
```

```{r, results='asis'}
blank_lines(2)
```

#### Other Columns

```{r sv_noBND2_other, eval=!no_sv_found}
details::details(sv_detail_table(sv_noBND_other))
sv_noBND_otherdt
```

#### Many Genes

```{r sv_noBND_manygenes, eval=!no_sv_found}
details::details(sv_detail_table(sv_noBND_manygenes))
sv_noBND_manygenesdt
```

#### Many Transcripts

```{r sv_noBND_manytx, eval=!no_sv_found}
details::details(sv_detail_table(sv_noBND_manytx))
sv_noBND_manytxdt
```

## Copy Number Variants
The purity and ploidy estimator
[PURPLE](https://github.com/hartwigmedical/hmftools/tree/master/purity-ploidy-estimator)
is used to generate a copy number profile for the somatic sample.

```{r, results='asis'}
blank_lines(1)
```

### UMCCR Gene Somatic CNV Calls {#cnv-gene}

<details>
<summary>Description</summary>

PURPLE copy number alterations in the UMCCR Cancer Gene panel (~1,200 genes) - description
is from <https://github.com/hartwigmedical/hmftools/blob/master/purity-ploidy-estimator/README.md#gene-copy-number-file>

```{r purple_gene_cnv_description}
purple_cnv_som_gene$descr %>%
  dplyr::mutate(
    Column = kableExtra::cell_spec(Column, bold = TRUE)) %>%
  knitr::kable(escape = FALSE) %>%
  kableExtra::kable_paper(c("hover", "striped"), full_width = FALSE, position = "left") %>%
  kableExtra::scroll_box(height = "200px")
```

</details>

```{r purple_gene_cnv_tab}
purple_cnv_som_gene$tab %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                options = list(scroller = TRUE, scrollY = 400, scrollX = TRUE, autoWidth = TRUE, keys = TRUE,
                               buttons = c('csv', 'excel'), dom = 'Bfrtip')) %>%
  DT::formatCurrency(~ start + end, currency = "", interval = 3, mark = ",", digits = 0) %>%
  DT::formatRound(~ minCN + maxCN, digits = 1)
```

```{r, results='asis'}
blank_lines(2)
```

### Genome-wide Somatic CNV Segments {#cnv-segs}

<details>
<summary>Description</summary>

PURPLE outputs a file with the copy number profile of all contiguous segments
of the tumor sample:

PURPLE copy number profile of all (contiguous) segments of the tumor sample - description
is from <https://github.com/hartwigmedical/hmftools/blob/master/purity-ploidy-estimator/README.md#copy-number-file>


```{r purple_seg_description}
purple_cnv_som$descr %>%
  dplyr::mutate(
    Column = kableExtra::cell_spec(Column, bold = TRUE)) %>%
  knitr::kable(escape = FALSE) %>%
  kableExtra::kable_paper(c("hover", "striped"), full_width = FALSE, position = "left") %>%
  kableExtra::scroll_box(height = "200px")
```

</details>

```{r purple_cnv_som_tab}
purple_cnv_som$tab %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                options = list(scroller = TRUE, scrollY = 400, scrollX = TRUE, autoWidth = TRUE, keys = TRUE,
                               buttons = c('csv', 'excel'), dom = 'Bfrtip')) %>%
  DT::formatCurrency(~ Start + End, currency = "", interval = 3, mark = ",", digits = 0)
```

```{r, results='asis'}
blank_lines(2)
```

### Genome-wide __Germline__ CNV Segments

```{r purple_cnv_germ_tab}
purple_cnv_germ$tab %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                class = "cell-border display compact",
                rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                options = list(scroller = TRUE, scrollY = 400, scrollX = TRUE, autoWidth = TRUE, keys = TRUE,
                               buttons = c('csv', 'excel'), dom = 'Bfrtip')) %>%
  DT::formatCurrency(~ Start + End, currency = "", interval = 3, mark = ",", digits = 0)
```

```{r, results='asis'}
blank_lines(2)
```

### PURPLE Charts {.tabset .tabset-fade #purple-charts}
PURPLE generates charts for summarising tumor sample characteristics.
Description is from the [PURPLE docs](https://github.com/hartwigmedical/hmftools/tree/master/purity-ploidy-estimator#charts).

#### Purity/ploidy
The following 'sunrise' chart shows the range of scores of all examined solutions of
purity and ploidy. Crosshairs identify the best purity / ploidy solution.

```{r purple_plot_purityrange, out.width="40%"}
knitr::include_graphics(file.path(img_dir, paste0(params$batch_name, '.purity.range.png')))
```

```{r, results='asis'}
blank_lines(2)
```

#### Copy number / Minor allele ploidy
The following figures show the AMBER BAF count weighted distribution of
copy number and minor allele ploidy throughout the fitted segments.
Copy numbers are broken down by colour into their respective minor
allele ploidy (MAP) while the minor allele ploidy figure is broken
down by copy number.

```{r purple_plot_copynumber, out.width="40%"}
knitr::include_graphics(c(file.path(img_dir, paste0(params$batch_name, '.copynumber.png')),
                        file.path(img_dir, paste0(params$batch_name, '.map.png'))))
```

#### Rainfall
If a somatic variant VCF has been supplied, a figure will be produced
showing the somatic variant ploidy broken down by copy number.

```{r purple_plot_somatic, out.width="40%"}
knitr::include_graphics(file.path(img_dir, paste0(params$batch_name, '.somatic.png')))
```

#### Clonality
The following diagram illustrates the clonality model of a typical sample.

The top figure shows the histogram of somatic ploidy for all SNVs and INDELs in blue.
Superimposed are peaks in different colours fitted from the sample as described in the
docs while the black line shows the overall fitted ploidy distribution. Red filled
peaks are below the 0.85 subclonal threshold.

We can determine the likelihood of a variant being subclonal at
any given ploidy as shown in the bottom half of the figure.

```{r purple_plot_somaticclonality, out.width="50%"}
knitr::include_graphics(file.path(img_dir, paste0(params$batch_name, '.somatic.clonality.png')))
```

#### Segment
The contribution of each fitted segment to the final score of the best fit is shown in
the following figure. Each segment is divided into its major and minor allele ploidy.
The area of each circle shows the weight (AMBER baf count) of each segment.

```{r purple_plot_segment, out.width="40%"}
knitr::include_graphics(file.path(img_dir, paste0(params$batch_name, '.segment.png')))
```

## Oncoviruses {.tabset .tabset-fade #oncoviruses}

Oncoviruses and their integration sites. Viral sequences are obtained from the
<a href="https://gdc.cancer.gov/about-data/data-harmonization-and-generation/gdc-reference-files">GDC database</a>.
Host genes are reported if the integration site falls on a gene or at least before 100kbp of the gene start.

```{r oncoviruses_table}
present_viruses <- params$oncoviral_present_viruses
breakpoints_tsv <- params$oncoviral_breakpoints_tsv

if (!is.na(breakpoints_tsv) && breakpoints_tsv != "" && file.exists(breakpoints_tsv)) {
  breakpoints <- readr::read_tsv(breakpoints_tsv, col_names = T, col_types = "cciicicccc")
  breakpoints %>%
    dplyr::mutate(Coord = ifelse(is.na(end), start, str_c(start, "-", end))) %>%
    dplyr::select(Contig = contig, Coord, `Read support` = PAIR_COUNT,
                  `Disrupted genes` = DisruptedGenes, `Upstream genes` = UpstreamGenes,
                  `SV type` = svtype, ID, MATEID) %>%
    DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                  class = "cell-border display compact",
                  rownames = FALSE, extensions = c("Scroller", "Buttons", "KeyTable"),
                  options = list(scroller = TRUE, scrollY = 400, scrollX = TRUE, autoWidth = TRUE, keys = TRUE,
                                 buttons = c('csv', 'excel'), dom = 'Bfrtip')) %>%
    DT::formatCurrency(~ Coord, currency = "", interval = 3, mark = ",", digits = 0)
} else {
  if (!is.na(present_viruses) && present_viruses != "" && file.exists(present_viruses)) {
    viruses <- readr::read_file(present_viruses)
    if (str_length(viruses) > 0) {
      cat(str_c("Detected significant traces of content of the following viruses: ",
                viruses,
                ", however no evidence of viral integration into the host genome is observed."))
    } else {
      cat("No oncoviral content detected in this sample.")
    }
  } else {
    cat("Oncoviral detection was disabled in the umccrise run with `-E oncoviruses`.")
  }
}
```

```{r, results='asis'}
blank_lines(1)
```

## Addendum {.tabset .tabset-fade #addendum}

<a href="#top">Back to top</a>

### Conda Pkgs Main

```{r conda_list_main}
conda_pkgs <-
  readr::read_table2(params$conda_list,
                     col_names = c("env", "name", "version", "build", "channel"),
                     col_types = "ccccc") %>%
  dplyr::mutate(channel = ifelse(is.na(channel), "MISSING", channel),
                env = sub("env", "umccrise", env),
                env = sub("^umccrise_", "", env)) %>%
  dplyr::arrange(name) %>%
  dplyr::group_by(name, version, build, channel) %>%
  dplyr::summarise(envs = paste(env, collapse = ", "), .groups = "drop_last") %>%
  dplyr::ungroup()

# main pkgs
tibble::tibble(
  name = c(
    "umccrise", "r-gpgr", "r-base", "pcgr", "cpsr", "cacao",
    "hmftools-purple", "hmftools-amber", "hmftools-cobalt", "hmftools-sage",
    "multiqc", "multiqc-bcbio", "reference-data",
    "samtools", "bcftools", "htslib", "ngs_utils", "python", "pandoc"
  )
) %>%
  dplyr::left_join(conda_pkgs, by = "name") %>%
  knitr::kable(format = "html") %>%
  kableExtra::kable_paper(c("hover", "striped"), full_width = TRUE, position = "left") %>%
  kableExtra::column_spec(1, bold = TRUE) %>%
  kableExtra::scroll_box(height = "300px")
```

### Report Inputs

```{r report_inputs}
report_inputs <- dplyr::tibble(key = names(params), value = unlist(params))
gpgr::write_tsvjsongz(report_inputs, "report_inputs", result_outdir)

report_inputs %>%
  knitr::kable(format = "html") %>%
  kableExtra::kable_paper(c("hover", "striped"), full_width = FALSE, position = "left") %>%
  kableExtra::column_spec(1, bold = TRUE) %>%
  kableExtra::scroll_box(height = "200px")
```

### Conda Pkgs All

```{r conda_list_all}
conda_pkgs %>%
  DT::datatable(filter = list(position = "top", clear = FALSE, plain = TRUE),
                rownames = FALSE, extensions = c("Scroller", "Buttons"),
                options = list(scroller = TRUE, scrollX = TRUE, scrollY = 400,
                               buttons = c('csv'), dom = 'Bfrtip'))
```
